patches/api/0389-Add-extra-raytracing-api.patch

```diff
@@ -0,0 +1,282 @@
++   */
++  @NotNull
++  public RayTraceBuilder raySize(double raySize) {
++    this.raySize = Math.max(0, raySize);
++    return this;
++  }
++
++  /**
++   * Override the fluid collision mode for the raytrace.
++   * Default value is {@link FluidCollisionMode#NEVER}.
++   *
++   * @param mode the new mode
++   * @return the modified builder
++   */
++  @NotNull
++  public RayTraceBuilder fluids(@NotNull FluidCollisionMode mode) {
++    Preconditions.checkArgument(mode != null, "Mode cannot be null");
++    this.mode = mode;
++    return this;
++  }
++
++  /**
++   * Override whether the raytrace should ignore passable blocks (blocks that the player can move through).
++   * Default value is true.
++   *
++   * @param ignorePassable the new value
++   * @return the modified builder
++   */
++  @NotNull
++  public RayTraceBuilder ignorePassable(boolean ignorePassable) {
++    this.ignorePassable = ignorePassable;
++    return this;
++  }
++
++  /**
++   * Define a set of specific blocks the raytrace should ignore.
++   * Default value is an empty set.
++   *
++   * @param ignoreBlocks the new set of blocks to ignore
++   * @return the modified builder
++   */
++  @NotNull
++  public RayTraceBuilder ignore(@NotNull Collection<Block> ignoreBlocks) {
++    Preconditions.checkArgument(ignoreBlocks != null, "Collection cannot be null");
++    this.ignoreBlocks = Set.copyOf(ignoreBlocks);
++    return this;
++  }
++
++  /**
++   * Define a predicate of entities to filter when casting the raytrace. All other entities will be ignored.
++   * Default value is colliding with all entities.
++   *
++   * @param entityPredicate the new predicate
++   * @return the modified builder
++   */
++  @NotNull
++  public RayTraceBuilder filter(@NotNull Predicate<Entity> entityPredicate) {
++    Preconditions.checkArgument(entityPredicate != null, "Predicate cannot be null");
++    this.entityPredicate = entityPredicate;
++    return this;
++  }
++
++  /**
++   * Build and cast the raytrace checking only blocks.
++   *
++   * @param world the world to cast the raytrace in
++   * @return the result
++   */
++  @Nullable
++  public RayTraceResult blocks(@NotNull World world) {
++    Preconditions.checkArgument(world != null, "World cannot be null");
++    Location start = new Location(world, origin.getX(), origin.getY(), origin.getZ());
++    return world.rayTraceBlocks(start, direction, range, mode, ignorePassable, ignoreBlocks);
++  }
++
++  /**
++   * Build and cast the raytrace checking both blocks and entities.
++   *
++   * @param world the world to cast the raytrace in
++   * @return the result
++   */
++  @Nullable
++  public RayTraceResult entities(@NotNull World world) {
++    Preconditions.checkArgument(world != null, "World cannot be null");
++    Location start = new Location(world, origin.getX(), origin.getY(), origin.getZ());
++    return world.rayTrace(start, direction, range, mode, ignorePassable, raySize, entityPredicate, ignoreBlocks);
++  }
```

Comment:
I think these methods should probably be on World, and just take a `RayTraceQuery` object.

Something to also do in the future (not in this PR) would be to make all the other ray trace methods utilize builders

Reply:
I think it's still useful to have utility methods to cast the raytrace directly from the builder.
If I add the methods with the `RayTraceContext` arg on World should I replace the new overload methods or simply add them. It would be rather odd to "lock" new api functionality behind the method with the context parameter. But on the other hand could be confusing to have multiple ways (and maybe implementations) to perform the same thing.

Reply:
I generally think that a "Builder" should have one job, to build some other object type. Now the Paper API isn't consistent on that, see the ParticleBuilder

Reply:
Likewise, you could argue the context object has one job, storing the raytrace configuration.
I think it's fair to assume that whenever one builds a context they also want to cast a raytrace using it. Those helper methods would simply combine the build and cast in one call.

Reply:
To be honest, i disagree; simply keeping data is not a responsibility: this turns objects into C structs. I don't see any problems of having a rayTrace method here.